subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+	 */
+	if ((len > (fs->sector_size - 5 * ate_size)) || ((len > 0) && (data == NULL))) {
+		return -EINVAL;
+	}
+
+	/* find latest entry with same id */
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(id)];
+
+	if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+		goto no_cached_entry;
+	}
+#else
+	wlk_addr = fs->ate_wra;
+#endif
+	rd_addr = wlk_addr;
+
+	while (1) {
+		rd_addr = wlk_addr;
+		rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+		if (rc) {
+			return rc;
+		}
+		if (wlk_ate.id == id) {
+			/* read the ate cycle only when we change the sector
+			 * or if it is the first read
+			 */
+			rc = zms_get_cycle_on_sector_change(fs, rd_addr, previous_sector_num,
+							    &current_cycle);
+			if (rc) {
+				return rc;
+			}
+			if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+				prev_found = true;
+				break;
+			}
+			previous_sector_num = SECTOR_NUM(rd_addr);
+		}
+
+		if (wlk_addr == fs->ate_wra) {
+			break;
+		}
+	}
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+no_cached_entry:
+#endif
+	if (prev_found) {
+		/* previous entry found */
+		rd_addr &= ADDR_SECT_MASK;
+		if (len > 8) {
+			rd_addr += wlk_ate.offset;
+		}
+
+		if (len == 0) {
+			/* do not try to compare with empty data */
+			if (wlk_ate.len == 0U) {
+				/* skip delete entry as it is already the
+				 * last one
+				 */
+				return 0;
+			}
+		} else if (len == wlk_ate.len) {
+			/* do not try to compare if lengths are not equal */
+			/* compare the data and if equal return 0 */
+			if (len <= 8) {
+				rc = memcmp(&wlk_ate.data, data, len);
+				if (!rc) {
+					return 0;
+				}
+			} else {
+				rc = zms_flash_block_cmp(fs, rd_addr, data, len);
+				if (rc <= 0) {
+					return rc;
+				}
+			}
+		}
+	} else {
+		/* skip delete entry for non-existing entry */
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	/* calculate required space if the entry contains data */
+	if (data_size) {
+		/* Leave space for delete ate */
+		if (len > 8) {
+			required_space = data_size + ate_size;
+		} else {
+			required_space = ate_size;
+			/* if data is in the ATE we need to make sure that we leave the ATE at
+			 * address 0x0 of the sector empty (even for delete ATE). Otherwise, the
+			 * fs->ate_wra will be decremented after this write by ate_size and it
+			 * will overflow to the previous sector.
+			 */
+			if (!((fs->ate_wra - ate_size) & ADDR_OFFS_MASK)) {
+				required_space += ate_size;
+			}
+		}
+	}
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	gc_count = 0;
+	while (1) {
+		if (gc_count == fs->sector_count) {
+			/* gc'ed all sectors, no extra space will be created
+			 * by extra gc.
+			 */
+			rc = -ENOSPC;
+			goto end;
+		}
+
+		if (fs->ate_wra >= (fs->data_wra + required_space)) {
+			rc = zms_flash_write_entry(fs, id, data, len);
+			if (rc) {
+				goto end;
+			}
+			break;
+		}
+		rc = zms_sector_close(fs);
+		if (rc) {
+			goto end;
+		}
+		rc = zms_gc(fs);
+		if (rc) {
+			goto end;
+		}
```

Comment:
Here is some logging missing, It might be important to log info that the gc is failing.